admin/services/check-start.service.js

```diff
@@ -1,46 +1,89 @@
+  // generate Pins
+  // find out all previous check forms by pupil for the current check window
+  // allocate a check form to each pupil randomly (in memory)
+  // save all the check forms
+  // connection -> commit transaction
+  // on error -> rollback transaction
+
+  // Find the check window we are working in
+  const checkWindow = await checkWindowDataService.sqlFindOneCurrent()
+
+  // Update the pins for each pupil
+  await pinGenerationService.updatePupilPins(pupilIds)
+
+  // Create the check for each pupil
+  const checks = []
+  for (let id of pupilIds) {
+    const c = await checkStartService.initialisePupilCheck(id, checkWindow)
+    checks.push(c)
+  }
+  await checkDataService.sqlCreateBatch(checks)
+}
+
+// Return a new pupil check object
+checkStartService.initialisePupilCheck = async function (pupilId, checkWindow) {
+  const checkForm = await checkFormService.allocateCheckForm()
 
-const checkStartService = {
-  /**
-   *
-   * @param pupilId
-   * @return {Promise<*>} checkCode - UUID v4
-   */
-  startCheck: async function (pupilId) {
-    // Get the current checkWindow, throw an error if there is no window available
-    // const checkWindow = await checkWindowDataService.fetchCurrentCheckWindow()
-
-    // Allocate a checkForm to a pupil, or will throw an error
-    const checkForm = await checkFormService.allocateCheckForm()
-    const checkWindow = await checkWindowDataService.sqlFindOneCurrent()
-
-    // Generate a new CheckCode for this unique check
-    const checkCode = uuidv4()
-    // TODO: the hard coded values below are in place until check form and window move to SQL
-    const checkData = {
-      pupil_id: pupilId,
-      checkCode,
-      checkWindow_id: checkWindow.id,
-      checkForm_id: checkForm.id,
-      pupilLoginDate: moment.utc(),
-      startedAt: null
-    }
-
-    // Save the details to the `Check` table
-    await checkDataService.sqlCreate(checkData)
-
-    return Promise.resolve({
-      checkCode,
-      checkForm
-    })
+  if (!checkForm) {
+    throw new Error('CheckForm not allocated')
   }
+
+  const checkData = {
+    pupil_id: pupilId,
+    checkWindow_id: checkWindow.id,
+    checkForm_id: checkForm.id
+  }
+
+  // Save the details to the `Check` table
+  return checkData
+}
+
+/**
+ *
+ * @param pupilId
+ * @return {Promise<*>} checkCode - UUID v4
+ */
+checkStartService.pupilLogin = async function (pupilId) {
+  const check = await checkDataService.sqlFindOneForPupilLogin(pupilId)
+  if (!check) {
+    throw new Error('Unable to find a prepared check for pupil: ' + pupilId)
+  }
+
+  const res = await checkFormDataService.sqlGetActiveForm(check.checkForm_id)
+  if (!res) {
+    throw new Error('CheckForm not found: ' + check.checkForm_id)
+  }
+  const checkForm = R.head(res)
+  const checkData = {
+    id: check.id,
+    pupilLoginDate: moment.utc()
+  }
+
+  await checkDataService.sqlUpdate(checkData)
+  const questions = JSON.parse(checkForm.formData)
```

Comment:
what happens if formData is undefined?
that will throw an error on parsing.
The error will be prevented if you do this instead:
`const questions =  checkForm.formData && JSON.parse(checkForm.formData)`

Reply:
checkForm.formData is non nullable

Reply:
fair enough